### PR TITLE
Phoneme Detection and Classifier Model Codes

### PR DESCRIPTION
**Pull request by Anirudh0707**

### CODE REVIEW

File: applications/KWS_Phoneme/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Phoneme based Keyword Spotting(KWS)
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:1]: Can we make it `Phoneme-based` here and in other places as well?

File: applications/KWS_Phoneme/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+# Project Description
+There are two major issues in the existing KWS systems (a) they are not robust to heavy background noise and random utterances, and (b) they require collecting a lot of data, hampering the ease of adding a new keyword. Tackling these issues from a different perspective, we propose a new two staged scheme with a model for predicting phonemes which are in turn used for phoneme based keyword classification. 
+
+First we train a phoneme classification model which gives the phoneme transcription of the input speech snippet. For training this phoneme classifier, we use a large public speech dataset like LibriSpeech. The public dataset can be aligned (meaning get the phoneme labels for each speech snippet in the data) using Montreal Forced Aligner. We also add reverberations and additive noise to the speech samples from the public dataset to make the phoneme classifier training robust to various accents, background noise and varied environment. In this project, we predict phonemes at every 10ms which is the standard way. You can find the aligned LibriSpeech dataset we used for training here.
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:6]: `meaning get` should be changed to `meaning we can get`. `environment` should be changed to `environments`.

File: applications/KWS_Phoneme/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+In the second part, we use the predicted phoneme outputs from the phoneme classifier for predicting the input keyword. We train a 1 layer FastGRNN classifier to predict the keyword based on the phoneme transcription as input. Since the phoneme classifier training has been done to account for diverse accent, background noise and environments, the keyword classifier can be trained using a small number of Text-To-Speech(TTS) samples generated using any standard TTS api from cloud services like Azure, Google Cloud or AWS.
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:8]: Change `diverse accent` to `diverse accents` here please. `api` should be changed to `API`.

File: applications/KWS_Phoneme/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Phoneme based Keyword Spotting(KWS)
+
+# Project Description
+There are two major issues in the existing KWS systems (a) they are not robust to heavy background noise and random utterances, and (b) they require collecting a lot of data, hampering the ease of adding a new keyword. Tackling these issues from a different perspective, we propose a new two staged scheme with a model for predicting phonemes which are in turn used for phoneme based keyword classification.
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:4]: Please capitalise `they` at the start of `a)` and `b)`

File: applications/KWS_Phoneme/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+1) Train a phoneme classification model on some public speech dataset like LibriSpeech
+2) Training speech dataset can be labelled using Montreal Force Aligner
+3) Speech snippets are convolved with reverberation files, and additive noises from YouTube or other open source are added
+4) We also add white gaussian noise of various SNRs
+
+# Training the KWS Model
+1) Our method takes as input the speech snippet and passes it through the phoneme classifier
+2) Keywords are detected by training a keyword classifier over the detected phonemes
+3) For training the keyword classifier, we use Azure and Google Text-To-Speech API to get the training data (keyword snippets)
+4) For example, if you want to train a Keyword classifier for the keywords in the Google30 dataset, generate TTS samples from the Azure/Google-Cloud/AWS API for each of the 30 keywords. The TTS samples for each keyword must be stored in a separate folder named according to the keyword. More details about how the generated TTS data should be stored are mentioned below in sample use case for classifier model training.
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:22]: Can we have full stops at the end of each point please? In this section as well as the above section? Also, `Keyword classifier` should be changed to `Keyword Classifier` or `keyword classifier`.

File: applications/KWS_Phoneme/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+## Phoneme Model Training
+The following command can be used to instantiate and train the phoneme model.
+```
+python train_phoneme.py --base_path=/path/to/librispeech_data/ --rir_base_path=/path/to/reverb_files/ --additive_base_path=/path/to/additive_noises/ --snr_samples="0,5,10,25,100,100" --rir_chance=0.5 
+```
+Some important command line arguments:
+1) base_path : Path of the speech data folder. The data in this folder should be in accordance to the dataloader code written here. 
+2) rir_base_path, additive_base_path : Path to the reverb and additive noise files
+3) snr_samples : List of various SNRs at which the additive noise is to be added.
+4) rir_chance : Probability at which reverberation has to be done for each speech sample
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:35]: What is the meaning of `Probability at which reverberation has to be done`? Also, please make use of full stops everywhere.

File: applications/KWS_Phoneme/README.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+## Classifier Model Training
+The following command can be used to instantiate and train the classifier model.
+```
+python train_classifier.py --base_path=/path/to/train_and_test_data_folders/ --train_data_folders=google30_azure_tts,google30_google_tts --test_data_folders=google30_test --phoneme_model_load_ckpt=/path/to/checkpoint/x.pt --rir_base_path=/mnt/reverb_noise_sampled/ --additive_base_path=/mnt/add_noises_sampled/ --synth 
+```
+Some important command line arguments:
+
+1) base_path : path to train and test data folders
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/README.md:44]: Please capitalize `path` here, and use proper punctuation.

File: applications/KWS_Phoneme/auxiliary_files/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+python download_youtube_data.py --csv_file=/path/to/csv_file.csv --target_folder=/path/to/target/folder/
+```
+
+Please check [Google's Audioset data page](https://research.google.com/audioset/download.html) for further details.
+
+The downloaded files would need to be converted to 16KHz for our pipeline. Please run the following for the same :
+```
+python convert_sampling_rate.py --source_folder=/path/to/csv_file.csv --target_folder=/path/to/target/16KHz_folder/ --fs=16000 --log_rate=100
+```
+The script can convert the sampling rate of any wav file to the specified --fs. But for our applications, we use 16KHz only.<br/>
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/auxiliary_files/README.md:20]: Please change `wav` to `.wav` and remove the HTML tag at the end of the sentence.

File: applications/KWS_Phoneme/auxiliary_files/README.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+Choose the log rate for how often the log should be printed for the sample rate conversion. This will print a string ever log_rate iterations.
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/auxiliary_files/README.md:21]: Please change `ever` to `every`.

File: applications/KWS_Phoneme/train_classifier.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                model['test_stats']['correct'] += float(np.sum((test_idx_pred == test_idx_label).detach().cpu().numpy()))
+                model['test_stats']['total'] += test_idx_label.shape[0]
+
+            test_accuracy = 100.0 * model['test_stats']['correct'] / model['test_stats']['total']
+            print(f"TEST => Accuracy: {test_accuracy}; Correct: {model['test_stats']['correct']}; Total: {model['test_stats']['total']}", flush=True)
+    return
+
+if __name__ == '__main__':
+    args = parseArgs()
+    train_classifier_model(args)
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/train_classifier.py:296]: Add a newline here please.

File: applications/KWS_Phoneme/train_phoneme.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        avg_ce = model['train_stats']['loss'].cpu() / total_batches
+        avg_err = 100 - 100.0 * (correct_frames / valid_frames)
+
+        print(f"Summary for Epoch {epoch} for Model {model['name']}", flush=True)
+        print(f"CE: {avg_ce}, ERR: {avg_err}, FRAMES {correct_frames} / {valid_frames}, PREDSTD: {epoch_prestd / total_batches}", flush=True)
+    return
+
+if __name__ == '__main__':
+    args = parseArgs()
+    train_phoneme_model(args)
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/train_phoneme.py:212]: Please add a newline here.

**Review by ShikharJ — COMMENTED** [review]
Minor Comments.

**Comment by Anirudh0707** [comment]
I apologize for the delay. I have incorporated all the typographical and punctuational changes in the new commit.

File: applications/KWS_Phoneme/auxiliary_files/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+python download_youtube_data.py --csv_file=/path/to/csv_file.csv --target_folder=/path/to/target/folder/
+```
+
+Please check [Google's Audioset data page](https://research.google.com/audioset/download.html) for further details.
+
+The downloaded files would need to be converted to 16KHz for our pipeline. Please run the following for the same :
+```
+python convert_sampling_rate.py --source_folder=/path/to/csv_file.csv --target_folder=/path/to/target/16KHz_folder/ --fs=16000 --log_rate=100
+```
+The script can convert the sampling rate of any .wav file to the specified --fs. But for our applications, we use 16KHz only.<br/>
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/auxiliary_files/README.md:20]: Please remove the HTML tag at the end of the sentence.

File: applications/KWS_Phoneme/auxiliary_files/README.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# Auxiliary Files to help Download and Prepare the Data
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/auxiliary_files/README.md:1]: Can we improve this title a bit, please?

File: applications/KWS_Phoneme/kwscnn.py [hunk 14]
```diff
@@ -0,0 +1,677 @@
...
+from torch.nn.parameter import Parameter
+from utils import _single
+import math
+import numpy as np 
+from edgeml_pytorch.graph.rnn import *
+
+class _IndexSelect(torch.nn.Module):
+    def __init__(self, channels, direction, groups):
+        """
+        Channel permutation module. The purpose of this is to allow mixing across the CNN groups
```
Comment (ShikharJ) [inline applications/KWS_Phoneme/kwscnn.py:17]: If you're using full-stops once in a statement, please use them throughout that statement.

**Comment by ShikharJ** [comment]
@harsha-simhadri I'm not the right person to review the Python scripts, so if Sachin and Anirudh feel they are fine, then I guess we should get them merged.
